lib/android-helpers.js

```diff
@@ -474,8 +479,18 @@ helpers.unlockWithHelperApp = async function (adb) {
     flags: "0x10200000",
     stopApp: false
   };
-  await adb.startApp(startOpts);
-  await adb.startApp(startOpts);
+
+  // Unlock succeed with a couple of retries.
+  await retryInterval(5, 1000, async function () {
+    logger.info("Unlocking screen");
+
+    // The command takes too much time so we should not call the command over twice continuously.
+    await adb.startApp(startOpts);
```

Comment:
I'm still not sure whether `waitDuration` argument should be at the default 20 seconds value here. Try to set it to 3-5 seconds, for example. We anyway only care about the result of `isScreenLocked` verification

Reply:
I read the code but the `waitDuration` work only we also set `waitActivity`.
```js
    if (startAppOptions.waitActivity) {
      await this.waitForActivity(startAppOptions.waitPkg, startAppOptions.waitActivity,
                                 startAppOptions.waitDuration);
```

But in this case, unlock app can't detect what `activity` should it wait since the command doesn't know what `activity` should be focused after unlocking. And the unlock app stops immediately after launching. (I didn't detect the activity using `dumpsys window` yet though, but we probably can detect it if we implement some scraping.)

And I tried `bluebird` timeout but the command continue to work in background. So, sometimes suddenly unlock happens by background commands or we can see the below errors for the commands.

```
Unhandled rejection Error: Cannot start the 'io.appium.unlock' application. Original error: Error executing adbExec. Original error: 'Command '/Users/kazuaki/Library/Android/sdk/platform-tools/adb -P 5037 -s emulator-5554 shell am start -W -n io.appium.unlock/.Unlock -a android.intent.action.MAIN -c android.intent.category.LAUNCHER -f 0x10200000' timed out after 20000ms'; Stderr: ''; Code: 'null'
    at ADB.callee$0$0$ (/Users/kazuaki/GitHub/appium/node_modules/appium-android-driver/node_modules/appium-adb/lib/tools/apk-utils.js:125:11)
    at tryCatch (/Users/kazuaki/GitHub/appium/node_modules/appium-android-driver/node_modules/babel-runtime/regenerator/runtime.js:67:40)
    at GeneratorFunctionPrototype.invoke [as _invoke] (/Users/kazuaki/GitHub/appium/node_modules/appium-android-driver/node_modules/babel-runtime/regenerator/runtime.js:315:22)
    at GeneratorFunctionPrototype.prototype.(anonymous function) [as throw] (/Users/kazuaki/GitHub/appium/node_modules/appium-android-driver/node_modules/babel-runtime/regenerator/runtime.js:100:21)
    at GeneratorFunctionPrototype.invoke (/Users/kazuaki/GitHub/appium/node_modules/appium-android-driver/node_modules/babel-runtime/regenerator/runtime.js:136:37)
    at <anonymous>
```

----

And when I observe the behaviour on emulator, it often took more than 5 sec ...

Reply:
thanks for the explanation, I got the issue.
I'll add the `timeout` option to startApp call, so the the internal `shell` command invocation will not block for so long

Reply:
what if we try something like this?

```js
    await retry(2, async function () {
     if (!(await adb.isScreenLocked())) {
       return;
     }

     logger.info("Unlocking screen");
 
     // The command takes too much time so we should not call the command over twice continuously.
     adb.startApp(startOpts);
     await waitForCondition(async () => !(await this.isScreenLocked()), {
      waitMs: 5000,
      intervalMs: 1000,
    });
   });
```

Reply:
I compared a couple of times, but the time was almost same...